target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,750 @@
+ *   #address-cells = <1>;
+ *   #size-cells = <1>;
+ *   ranges = <0 0x9f000000 0x18000>;
+ * };
+ *
+ * osc: oscillator {
+ *   compatible = "fixed-clock";
+ *   #clock-cells = <0>;
+ *   clock-frequency = <25000000>;
+ * };
+ *
+ * ccu: clock-controller {
+ *   compatible = "realtek,rtl8380-clock";
+ *   #clock-cells = <1>;
+ *   clocks = <&osc>;
+ *   clock-names = "ref_clk";
+ *   cpu-range = <MIN MAX>;
+ *   mem-range = <MIN MAX>;
+ *   lxb-range = <MIN MAX>;
+ * };
+ *
+ * The SRAM part is needed to be able to set clocks. When changing clocks the code must not run
+ * from DRAM. Otherwise system might freeze. Take care to adjust CCU compatibility, SRAM address
+ * and size to the target SOC device. To be able to use dynamic clock frequencies the range
+ * parameters must be supplied. If ommited the driver will only provide startup clock speeds.
+ * Afterwards one can access/identify the clocks in the other DT devices with <&ccu CLK_CPU>,
+ * <&ccu CLK_MEM> or <&ccu CLK_LXB>. Additionally the clocks can be used inside the kernel with
+ *
+ * cpu_clk = clk_get(NULL, "cpu_clk");
+ * mem_clk = clk_get(NULL, "mem_clk");
+ * lxb_clk = clk_get(NULL, "lxb_clk");
+ *
+ */
+
+#include <asm/cacheflush.h>
+#include <asm/mipsmtregs.h>
+#include <linux/clk.h>
+#include <linux/clk-provider.h>
+#include <linux/clkdev.h>
+#include <linux/cpu.h>
+#include <linux/delay.h>
+#include <linux/genalloc.h>
+#include <linux/io.h>
+#include <linux/ioport.h>
+#include <linux/of_address.h>
+#include <linux/of_platform.h>
+#include <linux/platform_device.h>
+#include <linux/slab.h>
+
+#include <dt-bindings/clock/rtl83xx-clk.h>
+#include "clk-rtl83xx.h"
+
+#define read_sw(reg)		readl(((volatile void *)RTL_SW_CORE_BASE) + reg)
+#define read_soc(reg)		readl(((volatile void *)RTL_SOC_BASE) + reg)
+
+#define write_sw(val, reg)	writel(val, ((volatile void *)RTL_SW_CORE_BASE) + reg)
+#define write_soc(val, reg)	writel(val, ((volatile void *)RTL_SOC_BASE) + reg)
+
+/*
+ * some hardware specific definitions
+ */
+
+#define SOC_RTL838X		0
+#define SOC_RTL839X		1
+#define SOC_COUNT		2
+
+#define MEM_DDR1		1
+#define MEM_DDR2		2
+#define MEM_DDR3		3
+
+#define REG_CTRL0		0
+#define REG_CTRL1		1
+#define REG_COUNT		2
+
+#define OSC_RATE		25000000
+
+static const int rtcl_regs[SOC_COUNT][REG_COUNT][CLK_COUNT] = {
+	{
+		{ RTL838X_PLL_CPU_CTRL0, RTL838X_PLL_MEM_CTRL0, RTL838X_PLL_LXB_CTRL0 },
+		{ RTL838X_PLL_CPU_CTRL1, RTL838X_PLL_MEM_CTRL1, RTL838X_PLL_LXB_CTRL1 },
+	}, {
+		{ RTL839X_PLL_CPU_CTRL0, RTL839X_PLL_MEM_CTRL0, RTL839X_PLL_LXB_CTRL0 },
+		{ RTL839X_PLL_CPU_CTRL1, RTL839X_PLL_MEM_CTRL1, RTL839X_PLL_LXB_CTRL1 },
+	}
+};
+
+#define REG_SET(_rate, _ctrl0, _ctrl1)		\
+	{					\
+		.rate = _rate,			\
+		.ctrl0 = _ctrl0,		\
+		.ctrl1 = _ctrl1,		\
+	}
+
+struct rtcl_reg_set {
+	unsigned int rate;
+	unsigned int ctrl0;
+	unsigned int ctrl1;
+};
+
+static const struct rtcl_reg_set rtcl_838x_cpu_reg_set[] = {
+	REG_SET(300000000, 0x045c8, 0x1414530e),
```

Comment:
do we have _any_ idea what any of these bitfields mean? I doubt it right? maybe add a comment that you pulled these ad-verbatim from the SDK and we do not know what these fields mean?

Reply:
of course a big idea: [look at WIKI](https://svanheule.net/realtek/maple/register/pll_cpu_ctrl0)